New rule: Auto-updating text content can be paused or stopped (#960)
act-rules/act-rules.github.io: _rules/auto-update-text-efbfc7.md

---
id: efbfc7
name: Text content that updates automatically can be paused, stopped or hidden
rule_type: atomic
description: |
  This rule checks that there are mechanisms to pause, stop or hide the auto-updating of text content.
accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgements:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [visible][] [HTML element][] in an [HTML document][] if:

> - the `innerText` property of the [element][html element] changes without the user [activating][activate] or [focusing][focus] any [element][html element] in the same [HTML document][];

Reviewer: I'm still having an uneasy feeling about the causality, but I'm not entirely sure how to express it… (sorry looks like this is going to be a long rant…)

---
As far as I understand, part of the problem that 2.2.2 detects is that the auto-updates never end. It is not a one time event but a continuous thing that can be distracting. So, something like this
```html
<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
<body onload="setTimeout(change, 1000)">
  <p>Random number: <span id="target">1</span></p>
</body>
```
should be good because there is only 1 update. There is no need to have a mechanism to stop the auto-updates because they stop themselves. This is currently applicable (content changes without user activating or focusing anything) and I'd say it fails the rule (there is no UI component to stop the change).

The first point that raises is that there is no real difference between an auto-update every second and a single update after 1, 2, 3, …, 3600 seconds… (`onload="for(var i=1; i<=3600; i++) { setTimeout(change, i*1000) }"`) There is something on the SC which is about the global perception (all these updates are part of the same auto-update, not individual events). That does not bode well for making an objective applicability…

The "auto-update" part of the SC does not have any time limit (contrarily to the "blinking" part which is OK if less than 5 seconds). But maybe we need to enforce it to make the applicability objective ("the time between the first and last (if any) change of the content is 5 seconds or more")?

---
Then, there is the causality between the click (or other user action) and the changes far in the future. If I look at Inapplicable example 2, I have no doubt that this is OK for WCAG. But the click is going to trigger updates far in the future. The 3600th update, happening 1 hour from now, may well look like it happens without the user activating an element. That is going to be OK because we are going to perceive all the updates as a single event and thus keep the causality in mind.

But if the updates are considered individually, it won't be that obvious that they happen due to that click. That hints that the "user activating" part is not really objective and unambiguous.

```html
<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
<p>Random number: <span id="target">1</span></p>
<input type="button" id="control" onclick="setTimeout(startUpdates, 5*60*1000)" value="Click me if you dare!" />
```
I think that is OK for WCAG (the updates start due to user interaction). It should also be inapplicable for the rle (the updates start on user activating an element). But I am not sure that the perception is going to be that the updates are linked to the click (because of the 5 minutes delay before it). And if I feel that the updates are not linked to the click, then this is applicable and fails the rule (but passes WCAG).

Compare with:
```html
<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
<body onload="setTimeout(startUpdates, 5*60*1000)">
  <p>Random number: <span id="target">1</span></p>
  <input type="button" id="control" value="Click me if you dare!" />
</body>
```
this second case is bad for WCAG and for the rule.
But if I don't have access to the code (or if code is obfuscated or the triggers hidden behind undecidable conditions), the difference between the two is very slim. Sure, I can wait in the first case and make enough experiences to convince me that the updates start as a consequence of me clicking the button. But that is not obvious and might be dangerously close to subjectivity (which is not allowed in Applicability).

---
What I'm trying to figure out here is how to make both the "action => updates" causality, and the "all these updates are a single thingie" unambiguous and objective.

Author: I guess I'm not reading the SC the same way you do:

- The SC does not apply only to auto-updates that never end. One example of an auto-updating content in the Understanding document is an "auto-advancing presentation". I certainly expect a presentation to end sometime :-)

- My interpretation of the SC is that it aims at making sure that users have the time they need to perceive and understand the content before it automatically updates (that is why a solution that changes the frequency of the updates also passes the SC - so we should take that into account, I guess) @

> - the `innerText` property of the [element][html element] changes without the user [activating][activate] or [focusing][focus] any [element][html element] in the same [HTML document][];

Author: Good catch. I'm redoing the "action" part.

- the [element][html element] does not have [children][child] whose `innerText` property is also changed;
- the change happens after the [readiness][document readiness] of the [HTML document][] the [element][html element] belongs to is equal to "complete";
- it is not the only [content][] in the [HTML document][].

## Expectation

> For the test target a [user interface component][] is provided to pause, stop or hide the change of the [text content][].

Reviewer: Should that UI component be visible and in the accessibility tree and with a non empty name?
(will need corresponding failed examples)

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [Understanding Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]
- [G186: Using a control in the Web page that stops moving, blinking, or auto-updating content][g186]
- [F16: Failure of Success Criterion 2.2.2 due to including scrolling content where movement is not essential to the activity without also including a mechanism to pause and restart the content][f16]

## Test Cases

### Passed

#### Passed Example 1

The text content automatically updates after the page completes loading. A button is available to stop the automatic updates.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>
	<input type="button" onclick="stopUpdates()" value="Stop updates" />

	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.

> 	</p>

Reviewer: I am not sure this is actually necessarily (in any of the examples).
There is already the "Random number:" part which makes the auto-updating `target` **not** the only content in the page. (and Inapplicable example 3 does not have it…)

Reviewer: This still stands. I think all examples can be trimmed of this `p` and made shorter and simpler.

	</p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Passed Example 3

The text content automatically updates after the page completes loading. A button is available to hide the automatically updating content.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>
	<input type="button" onclick="hide()" value="Hide updates" />

	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

### Failed

#### Failed Example 1

The text content automatically updates after the page completes loading. There is no component to stop or pause the automatic updates or to hide the updating content.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>

	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

> ### Inapplicable

Reviewer: For coverage of the  condition, we need inapplicable examples with:
- no auto-updating content;
- children whose content also changes;
- changes happening before the readiness is "complete".

For the children case, that might maybe not be inapplicable because then one of the children whose content changes is going to be applicable. So likely a passed example where the description says that the top-level element is not applicable because of its child changing, but the child is applicable and passing.

		implement accessibility.
	</p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

### Inapplicable

#### Inapplicable Example 1

The text content automatically updates every 3 seconds after the page completes loading but it is not visible.

```html
<body onload="startUpdates()">
	<p style="display: none">Random number: <span id="target">1</span></p>

	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Inapplicable Example 2

The text content automatically updates every 3 seconds but only as a result of a user interaction.

```html
<body>
	<p>Random number: <span id="target">1</span></p>
	<input type="button" id="control" onclick="startUpdates()" value="Start updates" />

	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Inapplicable Example 3

> The automatically updating text content is the only content in the document.

Reviewer: Technically, it's not the only content because there is also the button (and the "Random number:" text).

Author: 👍

</body>
```

### Failed

#### Failed Example 1

The text content of the `span` element automatically updates multiple times without user intervention. There is no [instrument][] to stop, pause, hide or alter the frequency of the automatic updates.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Failed Example 2

The text content of the `span` element automatically updates multiple times without user intervention. A button is available to stop the automatic updates, but the button only stops the updates while it is focused.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>
	<input type="button" onfocus="stopUpdates()" onblur="startUpdates()" value="Stop updates" />

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

### Inapplicable

#### Inapplicable Example 1

The text content automatically updates but only as a result of the user activating a button on the page.

```html
<body>
	<p>Random number: <span id="target">1</span></p>

> 	<input type="button" id="control" onclick="startUpdates()" value="Start updates" />

Reviewer 2: We should add a stop button to this; once this is started, I think we may need a mechanism to stop this.

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Inapplicable Example 2

The automatically updating text content is the only content in the document.

```html
<body onload="startUpdates()">

> 	<span id="target">1</span>

Reviewer 2: While yes, this is correct, I'd also argue that all of the other examples are cases where this counter is the only content on the page. A button and a "random number" label would definitely be part of the auto-updating content. I think the examples need to be improved here; have some other content on them to make them valid. I also think an additional assumption is necessary for this. While yes, this "other text" solution works a little, this shows quite well that it's not an air tight test method.

Author: I did have this additional content in the examples previously and it got cut out by reviews, because technically the label was other text. 
I do agree that having additional content, unrelated to the updating content, makes the examples better, and I'll add it again.

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Inapplicable Example 3

The document does not have text content that updates automatically.

```html
<body>
	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>
</body>
```

#### Inapplicable Example 4

> The color of the element is updated, but not its `innerText` property.

Reviewer 2: Interesting example... arguably this is blinking content, maybe? I think this should have a stop button on it. I don't think I want to get into the conversation of whether or not this satisfies the SC.

  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgements:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [visible][] [HTML element][] in an [HTML document][] if:

- **(not alone)** it is not the only [content][] in the [HTML document][]; and
- **(changed)** the `innerText` property of the [element][html element] changes; and
- **(no child changed)** the [element][html element] does not have [children][child] whose `innerText` property also changes; and
- the [document][html document] the [element][html element] belongs to:
  - **(ready)** has [readiness][document readiness] equal to "complete" before the change happens; and
  - **(inactive)** is associated with a [window][] that has the [last activation timestamp][] equal to "positive infinity".

## Expectation 1

For the test target there exists a [mechanism][] that might be used to:

- pause, stop or hide the change of the [visible text content][]; or
- alter the frequency of the changes of the [visible text content][].

> **Note**: If there is more than one test target, a single [mechanism][] may be used to pause, stop, hide or alter the frequency for all test targets.

Reviewer: I say this now need to be updated to follow more closely the wording of the definition ("mechanism to …")
```suggestion
For the test target there exists a [mechanism][] to achieve one of the following objectives:

- pause and resume the change of the [visible text content][]; or
- stop the change of the [visible text content][]; or
- hide the content that changes; or
- alter the frequency of the changes of the [visible text content][].

**Note**: If there is more than one test target, a single [mechanism][] to pause, stop, hide or alter the frequency for all test targets is sufficient.
```

It might also be worth to nail it further down in the example descriptions ("the button, together with the javascript function `foo`, form a mechanism to stop the updates").

rule_type: atomic
description: |
  This rule checks that there are mechanisms to pause, stop or hide the auto-updating of text content.
accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgements:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [visible][] [HTML element][] in an [HTML document][] if:

- **(not alone)** it is not the only [content][] in the [HTML document][]; and
- **(changed)** the `innerText` property of the [element][html element] changes; and
- **(no child changed)** the [element][html element] does not have [children][child] whose `innerText` property also changes; and
- the [document][html document] the [element][html element] belongs to:
  - **(ready)** has [readiness][document readiness] equal to "complete" before the change happens; and
  - **(inactive)** is associated with a [window][] that has the [last activation timestamp][] equal to "positive infinity".

## Expectation 1

> For the test target there exists a [mechanism][] to achieve one of the following objectives:

Author: Item 1 is taken care by the part of the applicability that states the document is inactive. If there was no user interaction and there was a change, then it had to be automatic.

Item 2 is not assumed and not even considered. If there is a single change (therefore not repeated), a user might still be prevented from perceiving the content that has been changed.

Item 3 is, in fact, being assumed. I'll take this into account, plus your comments on the mechanism definition and try to find a way to improve it.

---
id: efbfc7
name: Text content that updates automatically can be paused, stopped or hidden
rule_type: atomic
description: |
  This rule checks that there are instruments to pause, stop or hide the auto-updating of text content.
accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgments:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [visible][] [HTML element][] if:

> - **(changed)** the `innerText` property of the [element][html element] changes; and

Reviewer 2: This rule is still making a few assumptions that aren't accounted for:

- Just because an element is visible, doesn't mean the text is. For example you can have an element with a background image and text content that is clipped in some way. I saw that on a button a little while back. You can probably capture this in the applicability somehow. Perhaps just say the element has a visible text node?

- The rule is still assuming that any innerText change is going to repeat. I don't think that's true. Most JS frameworks set up their content after load. This would capture all of those events. Another drawback of the activation notification idea is that it only works on initial page loads. You can't test SPAs with this. It also has a pretty limited set of events, so you might accidentally miss things.

What I was thinking is doing something like this:

```suggestion
- **changed:** The `innerText` property of the [element][html element] changes multiple times within a 10 minute time span where there is no user interaction; and
```

You'll have to define user interaction. I think if we define that as any event that has isTrusted set to true. That would include most assistive technologies, perhaps except for browser extensions. We should have an assumption on that. A note should also be added about the 10 minutes. It's a completely arbitrary cut-off, we can just say that anything that updates less frequently, or something that doesn't update on a regular interval may still not satisfy the SC, but that testers generally don't wait hours to see if a page changes.

Lastly, can you update the style of these labelled lists? My suggestion above is using the format also used in WCAG. I think that's the way to go on these.

- **(no child changed)** the [element][html element] does not have [children][child] whose `innerText` property also changes; and

> - **(not alone)** the [element][html element] has [sibling][] or [ancestor][] elements with non-empty `innerText` properties whose value is different from the `innerText` of the test target; and

Reviewer 2: I like this a lot. Took me a few minutes to realise what this does though. Maybe add a note to explain this? I'm not sure this sibling thing is necessary though. If it has a sibling with a different innerText, it's parent will also have a different innerText.

input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgements:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [visible][] [HTML element][] in an [HTML document][] if:

- **(not alone)** it is not the only [content][] in the [HTML document][]; and
- **(changed)** the `innerText` property of the [element][html element] changes; and
- **(no child changed)** the [element][html element] does not have [children][child] whose `innerText` property also changes; and
- the [document][html document] the [element][html element] belongs to:
  - **(ready)** has [readiness][document readiness] equal to "complete" before the change happens; and
  - **(inactive)** is associated with a [window][] that has the [last activation timestamp][] equal to "positive infinity".

## Expectation 1

For the test target there exists a [mechanism][] that might be used to:

- pause, stop or hide the change of the [visible text content][]; or
- alter the frequency of the changes of the [visible text content][].

**Note**: If there is more than one test target, a single [mechanism][] may be used to pause, stop, hide or alter the frequency for all test targets.

## Expectation 2

> The existing [mechanism][] is [visible][], has an [accessible name][] that is not only [whitespace][], and is [included in the accessibility tree][].

Reviewer 3: Why should the mechanism be visible and included in the accessibility tree? That is not part of this SC is it?

Author: You are correct

> ## Assumptions

Reviewer 3: This rule needs an assumption that the changing text provides information.

Author: I'm not sure I agree with this one. 
One can argue that text is always information. But, even without arguing such, from the SC: "The intent of this Success Criterion is to avoid distracting users during their interaction with a Web page."
Even if the updating text does not provide information, it will (potentially) distract users. Thus, I would say that even that needs to be checked.

Reviewer 3: Text can definitely be decorative (think of random words that float somewhere in the background). While I agree that such floating words could distract the user and that the intent document mentions that users should not get distracted, I don't think distracting users is actually a WCAG violation. The SC clearly states that only **information** that gets updated needs to have such a mechanism.

However I don't like this interpretation either so I am willing to drop it even though I think it is technicality correct.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>
	<label for="interval">Content update frequency (seconds):</label>
	<input type="text" id="interval" />
	<input type="button" onclick="changeFrequency(document.getElementById('interval').value)" value="Change frequency" />

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

### Failed

#### Failed Example 1

The text content of the `span` element automatically updates after the page completes loading. There is no [mechanism][] to stop, pause, hide or alter the frequency of the automatic updates.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Failed Example 2

The text content of the `span` element automatically updates after the page completes loading. A button is available to stop the automatic updates, but the button is not visible.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>

> 	<input type="button" onclick="stopUpdates()" value="Stop updates" style="display:none" />

Reviewer 3: I would advice picking a different method of making this text invisible because using display:none also places the text outside of the accessibility tree.

> 	<input type="button" onclick="stopUpdates()" value="Stop updates" style="display:none" />

Reviewer 3: I would not fail this on this SC (but on other SC). I would advice to remove this example (and others)

---
id: efbfc7
name: Text content that updates automatically can be paused, stopped or hidden
rule_type: atomic
description: |
  This rule checks that there are instruments to pause, stop or hide the auto-updating of text content.
accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgments:
  authors:
    - Carlos Duarte
---

> ## Applicability

Reviewer 2: Sorry, i missed this before, but there are a couple of ambiguous DOM traversals here:
- with a visible text node <-- has a descendant in the flat tree
- does not have children <-- in the flat tree
- has an ancestor <-- in the flat tree

The rule applies to any [visible][] [HTML element][] in an [HTML document][] if:

- the `innerText` property of the [element][html element] changes;
- the [element][html element] does not have [children][child] whose `innerText` property is also changed;
- the change happens after the [readiness][document readiness] of the [HTML document][] the [element][html element] belongs to is equal to "complete";
- the change happens when the [last activation timestamp][] of the [window][] the [HTML document][] is associated with has the value "positive infinity";
- it is not the only [content][] in the [HTML document][].

## Expectation 1

> For the test target a [user interface component][] is provided to pause, stop, hide or alter the frequency of the changes of the [visible text content][].

Reviewer: If the stop mechanism is a global shortcut, this will fail the rule but pass the SC…

```html
<body onload="startUpdates()">
  <script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
  <p>Random number: <span id="target">1</span></p>
  <p>Press "+" to stop updates"</p>
  <script>
    document.body.addEventListener("keydown", function(event) {
      if (event.key === "+") { stopUpdates(); }
    }
  </script>
</body>
```

Author: You are right. The only way I can see to address this is to use [mechanism](https://www.w3.org/TR/WCAG21/#dfn-mechanism) similarly to what was done in [audio or video that plays automatically has a control mechanism](https://act-rules.github.io/rules/4c31df)

Reviewer: Yes. Problem is that we then need an unambiguous definition of mechanism...
#1016

> For each test target there exists an [instrument][] in the same [web page](#web-page-html), or in another [web page](#web-page-html) that is [linked][hyperlink] from the [web page](#web-page-html) of the test target, to achieve one of the following objectives:

Reviewer 2: Another issue, what if the instrument isn't in the page, but you can do something to get it there, such as opening a modal?

Author: A modal is still in the same web page. I've added a note to indicate the user might have to perform an action for the instrument to become available.

Reviewer 2: I disagree with that, which if nothing else suggests "in the same web page" is ambiguous. This needs to be fixed.

Author: Just to be clear: which part do you disagree with? That the modal is on the same web page? Or with the contents of the note?
(if it's the former, can you tell me why you disagree with it?)

Reviewer 2: I disagree that the content on a modal is in the same web page. It isn't until someone opens the modal.

- pause and resume the change of the [visible text content][]; or
- stop the change of the [visible text content][]; or
- hide the content that changes; or
- alter the frequency of the changes of the [visible text content][].

**Note:** If there is more than one test target, the same [instrument][] may be used to pause, stop, hide or alter the frequency of several or even all test targets.

## Assumptions

- This rule assumes that the auto-updating of the content is not [essential][], which is listed as valid exception to [Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]. When the auto-updating of content is [essential][] this rule may produce incorrect results.
- This rule assumes that any [content][] changes are enabled by the content of [HTML document][] the test target belongs to. Changes originated by any other sources (e.g. browser shortcuts, browser extensions, browser settings, user agents, external browser applications) are not considered.

> - This rule assumes that all user actions are transmitted by the user agent to the [HTML document][]. If there are other event sources that result from a user action this rule might fail but the success criterion might still pass.

Reviewer 2: success criteria don't "pass", they are "satisfied" (or not).

```suggestion
- This rule assumes that all user actions are transmitted by the user agent to the [HTML document][]. If there are other event sources that result from a user action this rule might fail but the success criterion might still be satisfied.
```

> - This rule assumes that all user actions are transmitted by the user agent to the [HTML document][]. If there are other event sources that result from a user action this rule might fail but the success criterion might still pass.
> - This rule assumes that available mechanisms for controlling the update of content rely on [activation][]. If there are other mechanisms that do not really on [activation][] the rule might fail but the success criterion might still pass.

Reviewer 2: ```suggestion
- This rule assumes that all user actions are transmitted by the user agent to the [HTML document][]. If there are other event sources that result from a user action this rule might fail but the success criterion might still pass.
- This rule assumes that available mechanisms for controlling the update of content rely on [activation][]. If there are other mechanisms that do not really on [activation][] the rule might fail but the success criterion might still be satisfied.
```

</body>
```

#### Passed Example 2

The text content of the `span` element automatically updates after the page completes loading. A button is available to pause and resume the automatic updates. The rule is not applicable to the first `p` element because it has a child (the `span` element) whose content updates.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>
	<input type="button" id="control" onclick="toggleUpdates()" value="Pause updates" />

	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Passed Example 3

The text content of the `span` element automatically updates after the page completes loading. A button is available to hide the automatically updating content. The rule is not applicable to the first `p` element because it has a child (the `span` element) whose content updates.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>

> 	<input type="button" onclick="hide()" value="Hide updates" />

Reviewer: ```suggestion
	<input type="button" onclick="hide()" value="Hide updating content" />
```

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>
	<input type="button" onclick="stopUpdates()" />

	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Failed Example 4

The text content of the `span` element automatically updates after the page completes loading. A button is available to stop the automatic updates, but the button is not included in the accessibility tree.

```html
<body onload="startUpdates()">
	<p>Random number: <span id="target">1</span></p>
	<input type="button" onclick="stopUpdates()" aria-hidden="true" />

	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

> ### Inapplicable

Reviewer: There should be an inapplicable example with no auto-updating content and/or one where the change is not on the `innerText` (e.g. color).

## Expectation

For each test target there exists an [instrument][] in the same [web page](#web-page-html), or in another [web page](#web-page-html) that is [linked][hyperlink] from the [web page](#web-page-html) of the test target, to achieve one of the following objectives:

- pause and resume the change of the [visible text content][]; or
- stop the change of the [visible text content][]; or
- hide the content that changes; or
- alter the frequency of the changes of the [visible text content][].

**Note:** If there is more than one test target, the same [instrument][] may be used to pause, stop, hide or alter the frequency of several or even all test targets.

**Note:** For the instrument to be available in the same web page, the user might have to perform an action (e.g. opening a modal dialog).

> ## Assumptions

Reviewer 2: There's another assumption here; that the text is actually "information". If you've got something like those command line spinners, or some animated progress bar using ASCII or whatever, I don't think that would fail the SC, but it does fail this rule.

Author: That is true. Added an assumption.

---
id: efbfc7

> name: Text content that updates automatically can be paused, stopped or hidden

Reviewer 4: Just a suggestion may not be necessary.
```suggestion
name: Autoupdating text content can be pause, stop or hide
```

rule_type: atomic
description: |
  This rule checks that for any text content that auto-updates in a 10 minute time span, there are instruments to pause, stop, or hide it or to control its update frequency. The arbitrary 10 minute time span, selected so that testing this rule would not be impractical, is not included in WCAG. Content that updates less frequently may fail success criteria 2.2.2 without failing this rule.
accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgments:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [HTML element][] that has a [visible][] [text node][] as a [descendant][] in the [flat tree][] if:

> - **changed:** the `innerText` property of the [element][html element] changes multiple times within a 10 minute time span where there is no [user interaction][]; and

Reviewer: Now that we've worked out (or will soon have) a definition of "changes in content" that includes a definition of "visible changes", I'm wondering whether we should upgrade this rule to use it (the "visible changes" part only, of course) 😕 

WDYT?

Author: @Jym77 that would imply a big change in the applicability (because "changes in content" applies to the web page, not specific elements). And, at least, it would make it difficult to specify that it is always the same content in the page that is being changed.

- **no child changed:** the [element][html element] does not have [children][child] in the [flat tree][] whose `innerText` property also changes; and
- **not alone:** the [element][html element] has an [ancestor][] element in the [flat tree][] with a non-empty `innerText` property whose value is different from the `innerText` of the test target.

## Expectation

For each test target there is at least one set of [instruments][instrument], where each [instrument][] is in the same [web page][] as the test target or can be found in a [clearly labeled location][] from that [web page][], to achieve one of the following objectives:

> - pause and resume the change of the [visible text content][]; or

Author: I've changed all uses of "update" to "changes" :-)

> - stop the change of the [visible text content][]; or

Reviewer 5: Or perhaps "stop the automated change of the visible text content". Presume this would be applicable to the pause/resume phrasing also.

> - stop the change of the [visible text content][]; or

Author: We don't use "automated" anywhere in the applicability or the expectation, so I don't think it's a good idea to introduce the term. Applicability specifies that the applicable changes are the ones that happen without user intervention (that's how we defined automated), and the changes the expectation addresses are those. Do you find this ambiguous?

> - stop the change of the [visible text content][]; or

Reviewer 5: I think I was agreeing more with the intent of what Adil suggested, rather than the exact wording. Automation may not be in the applicability or expectation, but it is referenced in the description.

I'm aware of set-ups where sports results might update automatically with Javascript, but with that feature disabled a page refresh would still update the content. I'm not sure if an adjustment is necessary, but the thing being paused or stopped is the automation of a change/update rather than there being no means to update the text. I understand if you don't feel that clarification is needed.

> - stop the change of the [visible text content][]; or

Author: But we already have an assumption for that :-)

"Any content changes are enabled by the content of the HTML document the test target belongs to. Changes originating from any other sources (e.g. browser shortcuts, browser extensions, browser settings, user agents, external browser applications) are not considered."
If the change is caused by the user refreshing the page, it is not considered for this rule.

> - stop the change of the [visible text content][]; or

Reviewer 5: If you think that is clear enough.

```html
<body onload="startUpdates()">
	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<p>Random number: <span id="target">1</span></p>

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Failed Example 2

This `span` element contains text content that automatically updates multiple times without user intervention and there is a button available to stop the automatic updates, but the button only stops the updates while it is focused.

```html
<body onload="startUpdates()">
	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<p>Random number: <span id="target">1</span></p>
	<input type="button" onfocus="stopUpdates()" onblur="startUpdates()" value="Stop updates" />

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

> ### Inapplicable

Author: @Jym77 There are several examples of elements that do not meet the **no child change** condition in the passed and failed examples. Those examples' descriptions point that out. I'm not really sure how I could come up with an example that would be fully inapplicable (because that child that changes would be applicable).

Good catch for the no text content example missing. I'm going to add it.

Also, I'll update descriptions and order.

rule_type: atomic
description: |
  This rule checks that for any auto-updating text content, there are instruments to pause, stop, or hide it or to control its update frequency.
accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgments:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [HTML element][] that has a [visible][] [text node][] as a [descendant][] in the [flat tree][] if:

- **changed:** the `innerText` property of the [element][html element] changes multiple times within a 10 minute time span where there is no [user interaction][]; and
- **no child changed:** the [element][html element] does not have [children][child] in the [flat tree][] whose `innerText` property also changes; and
- **not alone:** the [element][html element] has an [ancestor][] element in the [flat tree][] with a non-empty `innerText` property whose value is different from the `innerText` of the test target.

> **Note:** The 10 minute time span is an arbitrary limit which is not included in WCAG. Content that updates less frequently, or that doesn't update on a regular interval, will be inapplicable for this rule but may nonetheless fail [Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]. The accessibility problem tends to be less severe for longer time periods, and without a time limit, testing this rule consistently would be impractical.

Author: I've tried to do it. Please have a look.

## Expectation 1

> For each test target there exists a set of [instruments][instrument] to achieve one of the following objectives:

Reviewer 2: The word "exists" is still a little problematic here. This isn't something that can be defined. I realise you're limiting where it exists in expectation 2, but that still makes expectation 1 unanswerable. I think you may have to combine the two expectations in some way.

Reviewer: Yes, I agree that we probably need to close the set comprehension (i.e. combine with expectation 2) to avoid the following scenario:

1. Here's a target I've found on www.somewhere.org
2. Hey, here's an instrument to stop it, it's on www.elsewhere.com, so Expectation 1 passes.
3. Wait, there is absolutely nothing on www.somewhere.org that tells me to go to www.elsewhere.com, so the instrument is not in a clearly labeled location and Expectation 2 fails

Without combining, Expectation 1 is basically allowed to look anywhere on the web.

Author: I've suggested a possible combination of both expectations. Please review.

- pause and resume the change of the [visible text content][]; or
- stop the change of the [visible text content][]; or
- hide the content that changes; or
- alter the frequency of the changes of the [visible text content][].

**Note:** If there is more than one test target, the same [instrument][] may be used to pause, stop, hide or alter the frequency of several or even all test targets.

## Expectation 2

> For each [instrument][] in the set of [instruments][instrument] from Expectation 1, one of the following is true:
> 
> - the [instrument][] is in the same [web page](#web-page-html) of the test target; or
> - the [instrument][] can be found in a [clearly labeled location][].

Reviewer 2: I think inlining this is more readable:

```suggestion
Each [instrument][] in the set of [instruments][instrument] from Expectation 1 is either in the same [web page][] as the test target, or can be found in a [clearly labeled location][] from that [web page][].
```

> **Note:** If the set of instruments has more than one instrument, not every instrument of the set needs to be located in the same location.

Reviewer 2: I don't think this note is necessary.

**Note:** The 10 minute time span is an arbitrary limit which is not included in WCAG. Content that updates less frequently, or that doesn't update on a regular interval, will be inapplicable for this rule but may nonetheless fail [Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]. The accessibility problem tends to be less severe for longer time periods, and without a time limit, testing this rule consistently would be impractical.

## Expectation 1

For each test target there exists a set of [instruments][instrument] to achieve one of the following objectives:

- pause and resume the change of the [visible text content][]; or
- stop the change of the [visible text content][]; or
- hide the content that changes; or
- alter the frequency of the changes of the [visible text content][].

**Note:** If there is more than one test target, the same [instrument][] may be used to pause, stop, hide or alter the frequency of several or even all test targets.

## Expectation 2

For each [instrument][] in the set of [instruments][instrument] from Expectation 1, one of the following is true:

- the [instrument][] is in the same [web page](#web-page-html) of the test target; or

> - instructions for the user to locate the [instrument][] are in the same [web page](#web-page-html) of the test target.

Reviewer 2: Here's how far I got with an alternative. It definitely needs more work, taking out all the ambiguities, but this is roughly the direction I'm thinking:

> The instrument from expectation 1 can be found in a [clearly labelled location][]
> 
> # Clearly labelled location
> 
> Secondary information and alternative controls of functionality is often not displayed together with primary information or functionality. For example an option to change a web page to dark mode may be place on an options page, instead of being available in every page and page state of a website. Another example is a maps app, where instead of using GPS, an option is availble in a dropdown menu to set the current location of the device. Such content is (usually) placed in a clearly labelled location.
> 
> The location of some content can be said to be "clearly labelled" when it can be found by activating user interface components who's label indicates that the purpose of it may lead to finding the content, 
or to some further page or page state from which this action can be repeated until the content is found.
> 
> Whether or not the content is "clearly labelled" depends on the starting point of the search. If a page A has a clearly labelled link to some piece of important content, it is clearly labelled. Page B, which can be in the same website may not have such a link, or may have a link with an ambiguous link text, and so the location of the content starting from page B is not clearly labelled.

Author: @WilcoFiers new definition pushed, together with updated expectation. Can you please review again?

		implement accessibility.
	</p>

	<p>Random number: <span id="target">1</span></p>
	<input type="button" onclick="hide()" value="Hide updating content" />

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Passed Example 4

This `span` element has text content that automatically updates multiple times without user intervention and there is an [instrument][] available to change the update frequency. The rule is not applicable to the `p` element because it has a child (the `span` element) whose content updates.

```html
<body onload="startUpdates()">
	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<p>Random number: <span id="target">1</span></p>
	<label for="interval">Content update frequency (seconds):</label>
	<input type="text" id="interval" />
	<input type="button" onclick="changeFrequency(document.getElementById('interval').value)" value="Change frequency" />

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Passed Example 5

> This `span` element has text content that automatically updates multiple times without user intervention and there is an [instrument][] to control the updates available in the same window after the user activates the "Control updates" `button`.

Reviewer: This description now needs to be updated to reflect the new Expectation 2.

---
id: efbfc7
name: Text content that updates automatically can be paused, stopped or hidden
rule_type: atomic
description: |
  This rule checks that there are instruments to pause, stop or hide the auto-updating of text content.
accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgments:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [HTML element][] that has a [visible][] [text node][] as a [descendant][] in the [flat tree][] if:

> - **changed:** the `innerText` property of the [element][html element] changes multiple times within a 10 minute time span where there is no [user interaction][]; and

Author: - "update" does not encompass all the possible changes that might affect the innerText property. What if I remove the innerText? Would you say it was updated? In the very least, updated would be ambiguous then. But it changed.

- "based on preset interval" could imply that the updates have a determined frequency (because there is a preset interval). What if the updates are pushed by a server when a specific event occurs (which could be random, for all we know)? That is why multiple times was used.

- There is a note below explaining the "10 minutes". Don't you think it is enough? If we remove the arbitrary time span, it means that we would never be able to test the rule because one could never say that an update would not happen in the future (unless you inspected the scripts in all pages you are evaluating).

> - **changed:** the `innerText` property of the [element][html element] changes multiple times within a 10 minute time span where there is no [user interaction][]; and

Author: @adilsofficial let me begin by telling you that what you are suggesting is mostly how this rule looked like when I started writing it :-) Then, as you can see, it moved away from that after several rounds of reviews

- The "update" word. I think it was Wilco who raised this one. "Update" can be interpreted as you are interpreting (which was also my initial interpretation), but can also be interpreted as referring only to changes in the content of something (similar to what an update does in a database). If the "update" involves removing an element and re-creating the element with new content via a script, it should still be considered an update for this rule. But some people will not interpret that as an update (they will see it has a delete followed by a create, using the database terminology). That is why we replaced "update" by "change".

- The "preset interval". This one was raised by Jean-Yves, I believe. First, how do you determine that there is a preset interval (remember, this is applicability, so it needs to be objective)? Do you require the evaluator to read every possible script (including potentially server-side scripts) that is used on the web page to find that out? And what if the because the processor has a higher load, the processing of the event that updates the content takes an extra millisecond? Or the network is slower, and the update reaches the page one second after what it should? Then your updated frequency changes, and you cannot define a "preset interval". That is why, instead of trying to come up with an objective solution for this problem, we went with the "changes multiple times" solution.

You can find more detailed arguments for these issues in the history of comments for this PR :-)

Let us know if using "multiple times" in this rule is something that will prevent you from approving it (and if so, it would be great if you can come up with a way to solve the issue of the frequency/preset interval)

- **no child changed:** the [element][html element] does not have [children][child] in the [flat tree][] whose `innerText` property also changes; and
- **not alone:** the [element][html element] has an [ancestor][] element in the [flat tree][] with a non-empty `innerText` property whose value is different from the `innerText` of the test target.

> **Note:** The test targets for this rule are elements that have their text content change multiple times, on a page with further content, if that change happens without user intervention.

Author: If I remember this note was created before the applicability requirements were itemized. It was meant to provide a more understandable version, even if less precise. In fact it is not adding anything to the applicability, and I think it may be deleted.

> **Note:** The 10 minute time span is an arbitrary limit which is not included in WCAG. Content that updates less frequently, or that doesn't update on a regular interval, will be inapplicable for this rule but may nonetheless fail [Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]. It is however a much lesser accessibility issue, and without any time limit this rule would be virtually untestable.

Reviewer 2: ```suggestion
**Note:** The 10 minute time span is an arbitrary limit which is not included in WCAG. Content that updates less frequently, or that doesn't update on a regular interval, will be inapplicable for this rule but may nonetheless fail [Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]. The accessibility problem tends to be less severe for longer time periods, and without a time limit, testing this rule consistently would be impractical.
```

## Expectation 1

> For each test target there exists a set of [instruments][instrument] to achieve one of the following objectives:

Author: - The test-target is auto-updating (otherwise it wouldn't be a test target), therefore we don't need to repeat that.

- The instrument definition requires the objectives to be explicitly listed. That is what the list is: a set of objectives. It is not a set of requirements for the instruments.

> - pause and resume the change of the [visible text content][]; or

Author: Because the html element that is the test target can have more content that is not visible or text. If I changed the non-visible content, or the non-text content, I would be passing the rule, but failing the SC.

- stop the change of the [visible text content][]; or
- hide the content that changes; or

> - alter the frequency of the changes of the [visible text content][].

Author: Can you elaborate about what is ambiguous in "alter"?

> - alter the frequency of the changes of the [visible text content][].

Author: For consistency, I've replaced "alter" with "control"

> **Note:** If there is more than one test target, the same [instrument][] may be used to pause, stop, hide or alter the frequency of several or even all test targets.

Author: I think you are mixing the test targets with the instruments. The test targets are the content that is auto-updating. The instruments are the control mechanisms. What this note says is that I can have one single pause button to pause all auto-updating content in the page (or I can have one pause button for each auto-updating content). You seem to be implying that the note says the same button can be used to pause and stop the auto-updates. This is not what the note says.
Please correct me if I misunderstood your comment.

> **Note:** If there is more than one test target, the same [instrument][] may be used to pause, stop, hide or alter the frequency of several or even all test targets.

Reviewer 4: - Yes, I understood the same as you mentioned in the last sentence of your comment, that is the same button (instrument) can be used to pause, stop, and hide the auto-updates. if you meant here that "It can have one single pause button to pause all auto-updating content in the page (or I can have one pause button for each auto-updating content)." then please explain it a bit more to make it clear. because it seems that the same instrument may be used to perform all actions (pause, stop, and hide) of several or even all test targets.
 - Additional comment: Also, one more thing (maybe in assumptions) if you think is relevant and appropriate that is "it must be possible to use other contents on the page when content has been paused, stop, or hide." because it is not sufficient for content only to be paused when it is in focus (with the cursor or keyboard focus). The objective of avoiding distracting the user is not met in this case.

> **Note:** If there is more than one test target, the same [instrument][] may be used to pause, stop, hide or alter the frequency of several or even all test targets.

Author: - I've rephrased the note. What do you think now?

- That is okay since we don't claim to meet the SC when the rule passes. But I guess we can add another assumption to clarify that.

> ## Expectation 2

Author: The rules does not claim to have a pass-on-pass with the SC. So, even if the rule passes, further testing is needed. That would be one thing to test :-)

I don't think it would be easy to write an expectation according to "that does not tie up the user"... there are so many possible interpretations for this :-)

> Each [instrument][] in the set of [instruments][instrument] from Expectation 1 is in one of the following:

Reviewer 2: I don't think the approach you are taking here works (see comments below). I don't think it's practical to expect a tester to just allow this instrument to be effectively anywhere, and for a tester to have to go hunt around for it. If the homepage has auto-updating content, and the mechanism to turn it off is in the sitemap, I would have no problem failing that.

It seems to me that what we need here is a new concept. Something that can be paired with "instrument", describes where testers should go look for instruments. I'm not sure how that could be done, but I don't see a way for us to progress with this rule without it.

Author: Thanks for this feedback. It gave me an idea of what concept could be paired with "instrument". 
As you say, the tester should not have to go hunt for the "instrument". Neither should the user. 
Therefore, what I propose is that the instrument should be located in the same web page, or instructions on how to find the instrument should be located in the same web page.
I'm updating the expectation based on this (also of the other 3 rules that suffer from the same "instrument" related problem). Looking forward to you (quick) feedback.

> - the [instrument][] is in the same [web page](#web-page-html) of the test target; or

Author: It is not. The instrument can be on the same page of the auto-updating content *or* in another location, as long as the location is clearly labeled.

> - the same [web page](#web-page-html) of the test target as a result of an action from the user; or

Reviewer 2: I think this is simultaneously too restrictive and to open-ended. Too restrictive because it can only ever be a single "action" away. And too open-ended because "what is an action"? How do you test this consistently?

> - the [instrument][] can be found in a [clearly labeled location][].

Author: I can't find anywhere in the SC where it says that the instrument has to be next to the content or at the start of the page. I must be missing it. Could you please point me to it?

> - another [web page](#web-page-html) that is [linked][hyperlink] from the [web page](#web-page-html) of the test target.

Reviewer 2: This doesn't really work... what state is this page supposed to be in when the tester looks for the instrument? This still seems ambiguous.

- All user actions are transmitted by the user agent to the [HTML document][]. If there are other event sources that result from a user action this rule might fail but the success criterion might still be satisfied.
- Available mechanisms for controlling the update of content rely on [activation][]. If there are other mechanisms that do not really on [activation][] the rule might fail but the success criterion might still be satisfied.

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [Understanding Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]
- [G186: Using a control in the Web page that stops moving, blinking, or auto-updating content][g186]
- [F16: Failure of Success Criterion 2.2.2 due to including scrolling content where movement is not essential to the activity without also including a mechanism to pause and restart the content][f16]

## Test Cases

### Passed

#### Passed Example 1

> This `span` element has text content that automatically updates multiple times without user intervention and there is a button available to stop the automatic updates. The rule is not applicable to the `p` element because it has a child (the `span` element) whose content updates.

Author: Good catch. The description applies to the second `p` element. It became ambiguous when I added the first `p` element for the example to have more content. 
The `span` here is the only applicable element (that is why the example passes).
I'll update the description(s)!

		implement accessibility.
	</p>

	<p>Random number: <span id="target">1</span></p>
	<input type="button" onclick="hide()" value="Hide updating content" />

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Passed Example 4

This `span` element has text content that automatically updates multiple times without user intervention and there is an [instrument][] available to change the update frequency. The rule is not applicable to the `p` element because it has a child (the `span` element) whose content updates.

```html
<body onload="startUpdates()">
	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<p>Random number: <span id="target">1</span></p>
	<label for="interval">Content update frequency (seconds):</label>
	<input type="text" id="interval" />
	<input type="button" onclick="changeFrequency(document.getElementById('interval').value)" value="Change frequency" />

	<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
</body>
```

#### Passed Example 5

> This `span` element has text content that automatically updates multiple times without user intervention and the location of an [instrument][] to control the updates is clearly available.

Author: Why would you say this example "ties up the user or the focus"? 
1 - The page loads and content begins auto-updating (auto-updates on)
2 - The user activates the "Control updates" button (auto-updates on)
3 - The user activates the "Pause updates" buttons (auto-updates off)
4 - The user activates the "Dismiss" button (auto-updates off)
5 - The user uses the pages without any auto-updating content (auto-updates off)

So, how is the user prevented to use the page when the content has been paused?

```html
<body onload="startUpdates()">
	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<p>Random number: <span id="target">1</span></p>
	<p>To control the random number updates activate the "Control updates" button.</p>
	<input type="button" onclick="openModal()" value="Control updates" />

	<div
		style="
      display: none;
      position: fixed;
      top: 2em;
      left: 10em;
      background-color: #505050;
      color: white;
      padding: 1em;
      padding-top: 0em;
    "
		id="overlay"
	>
		<p>Control updates</p>
		<input type="button" id="control" onclick="toggleUpdates()" value="Pause updates" />
		<input type="button" onclick="hide()" value="Hide updates" />
		<button onclick="closeModal();">Dismiss</button>
	</div>

	<script type="text/javascript">
		function openModal() {
			document.getElementById('overlay').style.display = 'block'
		}
		function closeModal() {
			document.getElementById('overlay').style.display = 'none'
		}
	</script>
	<script type="text/javascript" src="test-assets/efbfc7/script.js"></script>
</body>
```

### Failed

#### Failed Example 1

> This `span` element has text content that automatically updates multiple times without user intervention and there is no [instrument][] available to stop, pause, hide or alter the frequency of the automatic updates.

Reviewer 4: ```suggestion
This `span` element has text content that automatically updates multiple times without user intervention and there is no [instrument][] available to stop, pause, hide or change the update frequency for auto-updating content.
```

---
id: efbfc7
name: Auto-updating text content can be paused or stopped
rule_type: atomic

description: |
  This rule checks that there are mechanisms to pause or stop the auto-updating of text content

accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree

authors:
  - Carlos Duarte
---

## Applicability

> The rule applies to any [text node][] that is [visible][] and a [child][] in the [flat tree][] of an HTML element, whose content changes through an action that was not initiated by the user, anytime after the [readiness][document readiness] of the document the [text node][] belongs to is equal to "complete".

Reviewer 6: Can we rephrase the applicability, it is a bit hard to read & comprehend.

Reviewer 6: When would be a case of a text node that is not a child of HTML element? Are you trying to exclude SVG text nodes?

Author: I tried to split it into items so that it is, hopefully, easier to read.
I've updated the applicability to consider any text node in the flat tree, so SVG text nodes are considered.

> The rule applies to any [text node][] that is [visible][] and a [child][] in the [flat tree][] of an HTML element, whose content changes through an action that was not initiated by the user, anytime after the [readiness][document readiness] of the document the [text node][] belongs to is equal to "complete".

Reviewer 6: what about nodes that are not part of the accessible tree?

Author: I might be misinterpreting the SC, but my understanding is that it applies to visible content, whether or not it is part of the accessibility tree. The understanding document states that:
- "Moving, blinking and scrolling" refers to content in which the visible content conveys a sense of motion
- The requirements for moving, blinking and scrolling content and for auto-updating content are the same (except for the possibility of allowing users to control the frequency of updates and the 5 second rule)
So my understanding is that the content must be visible, independent of being in the accessibility tree. 
What is everyone's opinion on this?

## Expectation 1

> For each test target a [mechanism][] is provided to pause or stop the updating of the content of the [text node][].

Reviewer 6: Also why not just do `[my-definition](link-to-my-definition)`, as against enumerating it all at the bottom of the time.

This pollutes the `spelling-ignore.yml` file to have keywords that are not necessarily english dictionary words, but made up hypenated words.

Author: Isn't that what the spelling-ignore is for? Anyway, if this is up for discussion, a new issue should be opened.

**Note**: If there is more than one [text node][] with auto-updating content, a single [mechanism][] may be used to pause or stop updating all [text nodes][text node].

## Expectation 2

The [mechanism][] to pause or stop the auto-updating of the content of the [text node][] is [visible][], has an [accessible name][] that is not empty ("") nor only [ASCII whitespace][] , and is [included in the accessibility tree][].

## Assumptions

> This rule assumes that the auto-updating of the content is not [essential][], which is listed as valid exception to SC 2.2.2.

Reviewer 6: I think if we are going to quote an SC, it should be linked.

> This rule assumes that the auto-updating of the content is not [essential][], which is listed as valid exception to SC 2.2.2.

Reviewer 7: See https://github.com/act-rules/act-rules.github.io/pull/960/files#r332913127.

Author: I've updated the assumption to make it clear that when the auto-updating is essential then the rule may produce incorrect results.

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

> - [Understanding Success Criterion 2.2.2: Pause, Stop, Hide](https://www.w3.org/WAI/WCAG21/Understanding/pause-stop-hide.html)

Reviewer 6: I would now argue that perhaps these should also be references at the bottom of the page, given we follow that. 

Please let's unify how hyperlinks are authored.

Author: moved the references to the bottom

- [G186: Using a control in the Web page that stops moving, blinking, or auto-updating content](https://www.w3.org/WAI/WCAG21/Techniques/general/G186)
- [F16: Failure of Success Criterion 2.2.2 due to including scrolling content where movement is not essential to the activity without also including a mechanism to pause and restart the content](https://www.w3.org/WAI/WCAG21/Techniques/failures/F16)

> ## Test Cases

Reviewer 6: Let us say I have an async fn that fetches data, in the meantime, the view says `loading`, then I also want to keep changing the text 
- `loading from skynet`
- `rebooting drives`
- `the sun is a big orange` etc., when the user is waiting

then finally after data get, I render

Is this considered auto updating text? In my opinion it is a status/ progress bar. 
All flight search websites do this today.

Reviewer 7: I would perhaps argue that such auto-updating text is essential, which leaves me questioning the assumption that it isn't.

Author: I tend to agree with Kasper. I would say that status bar is essential auto-updating text.

Reviewer 8: I agree that a status update bar would be considered essential auto-updating text.

SC 2.2.2 has this note: An animation that occurs as part of a preload phase or similar situation can be considered essential if interaction cannot occur during that phase for all users and if not indicating progress could confuse users or cause them to think that content was frozen or broken.

Author: As previously mentioned, I agree that this is an essential auto-update, therefore, as stated in the assumption and SC, there is no need to have a mechanism to control it.

## Background

- [Understanding Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]
- [G186: Using a control in the Web page that stops moving, blinking, or auto-updating content][g186]
- [F16: Failure of Success Criterion 2.2.2 due to including scrolling content where movement is not essential to the activity without also including a mechanism to pause and restart the content][f16]

## Test Cases

### Passed

#### Passed Example 1

The text content automatically updates every 3 seconds after the page completes loading. A button is available to stop the automatic updates.

```html
<body onload="start()">
	<p>Random number: <span id="target">1</span></p>
	<input type="button" onclick="stop()" value="Stop updates" />

	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		var updates
		function start() {
			updates = setInterval(change, 3000)
		}

		function stop() {
			clearInterval(updates)
		}

> 	</script>

Reviewer: Most of the Javascript is shared between example. For the sake of separation of concerns, DRY, and keeping the examples easier to focus on the important changes, I suggest:
* Move Javascript to a `test-assets` file.
* Put in it all the code: `change`, `start`, `toggle`, `stop`, `hide` and the `updating` boolean.
* (maybe rename to `startUpdates` and so on, maybe not…)
* Adapt `start` to also change `updating`.
* Systematically call `onload=start()` which makes purpose obvious. (I know I've been the one saying to put `onload=toggle()` at some places 😁 refactoring and polishing code is a very iterative process…)
* Only change the content of `onclick` in the individual examples. That will allow readers to focus on the important change and immediately agree with the description.

(Note that you can also decrease the update interval to 1s or 0.5s to make the behaviour of the examples even more obvious when trying them in the browser)

Author: Refactored

		implement accessibility.
	</p>

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		function start() {
			setInterval(change, 3000)
		}

		function hide() {
			var target = document.getElementById('target')
			target.style.visibility = 'hidden'
		}
	</script>
</body>
```

### Failed

#### Failed Example 1

The text content automatically updates every 3 seconds after the page completes loading. There is no component to stop or pause the automatic updates.

```html
<body onload="start()">
	<p>Random number: <span id="target">1</span></p>

	<p>
		The W3C Web Accessibility Initiative (WAI) develops standards and support materials to help you understand and
		implement accessibility.
	</p>

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		function start() {
			updates = setInterval(change, 3000)
		}
	</script>
</body>
```

#### Failed Example 2

> The text content (part of a progress bar) automatically updates every second after the page completes loading. There is no component for the user to pause the auto-updates, other than waiting for them to end.

Reviewer: I don't think that the core code for this example should be different from the others.
The same random number stuff can be used, no need for the "pick a price" stuff, just random number is OK.
The progress bar can probably be replaced by a timer displaying the remaining time. That should remove most or all of the styling and trim down the example.
Actually, the progress bar can probably get completely removed. It is only need because you have a button for stopping the random number, but if that button is removed, then the progress bar can be remove (there will be a random number auto updating for some time with no way to stop it, which is enough to fail the rule).

Author: I guess that this is a legacy test from a previous version of the rule where the progress bar made sense to have in a test. Given that the rule has been simplified, we don't even need this test anymore (if I followed your suggestions - with which I agree - we would end up with a test that was equivalent to failed example 1). I'm going to remove this.

```html
<html>
	<head>

> 		<title>Example page for rule efbfc7</title>

Reviewer: I'm not sure why only this example has a `title` (which is not really needed for the rule).

Author: It is the only one with `html` and `head` elements. The tests require a `title` when you have an `html` element, but not otherwise.

---
id: efbfc7
name: Text content that updates automatically can be paused, stopped or hidden
rule_type: atomic
description: |
  This rule checks that there are mechanisms to pause, stop or hide the auto-updating of text content
accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgements:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [visible][] [text content][] in an [HTML document][] if:

> - the [text content][] is removed or replaced by new content that does not contain the previous content; and

Reviewer 2: I don't think the definition of text content can be used here. Text content is computed from a node, it isn't a node in itself. What nodes should be tested here? Take this example:

```html
<div id="x"><b>Hello world</b></div>
<script>
setTimeout(() => document.querySelector('#x').innerHTML = '<i>Goodbye</i>', 1000)
</script>
```

If you mean to compute the text content of the `b` element, the text content isn't replace. The `b` element itself is. But for the `div`, the text content is replaced. You could maybe say the text content of the body, although I'm not sure how well that works. I'd prefer a clear definition of what "removed or replaced" means in that case.

Author: I'm not really understanding what the problem is, then. According to the applicability, the rule is not applicable to the text content of the `b` element and is applicable to the text content of the `div` element. 
From the user's perspective, there is textual content being replaced. From the rule's perspective, the applicability is catching that. 
@WilcoFiers do you still think it needs to be changed?

> - the removal or replacing is not the result of a [user intervention][]; and

Reviewer 2: I think this is much better. Thank you.

```suggestion
- the removal or replacing is not caused by a [user interaction][]; and
```

---
id: efbfc7
name: Auto-updating text content can be paused or stopped
rule_type: atomic

description: |
  This rule checks that there are mechanisms to pause or stop the auto-updating of text content

accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree

authors:
  - Carlos Duarte
---

## Applicability

> 

Reviewer 8: According to the SC, the auto-updating information (or text, as is the case of this test rule) must be presented in parallel with other content in order to be applicable. Work this into the applicability somehow?

Also add an inapplicable test case?

Author: Nice catch! I'll update accordingly.

The rule applies to any [visible][] [text node][] in the [flat tree][] of an [HTML document][] if:

- the content of the [text node][] changes through an action that was not initiated by the user; and
- the change happens anytime after the [readiness][document readiness] of the [HTML document][] the [text node][] belongs to is equal to "complete".

## Expectation 1

For the test target a [mechanism][] is provided to pause or stop the updating of the content of the [text node][].

**Note**: If there is more than one [text node][] with auto-updating content, a single [mechanism][] may be used to pause or stop updating all [text nodes][text node].

## Expectation 2

The [mechanism][] to pause or stop the auto-updating of the content of the [text node][] is [visible][], has an [accessible name][] that is not empty ("") nor only [ASCII whitespace][] , and is [included in the accessibility tree][].

> ## Assumptions

Reviewer: Maybe add that the rule assumes that the name of the mechanism does not need to be descriptive to pass the SC, and a non descriptive name is likely a violation of [some other SC].

Author: In reality, we are *not* assuming that the name is descriptive, so it is a "no assumption" :-)
Anyway, that can be helpful, so I'll add it as a note to expectation 2.

> This rule assumes that the auto-updating of the content is not [essential][], which is listed as valid exception to [SC 2.2.2][]. When the auto-updating of content is [essential][] this rule may produce incorrect results.

Reviewer: The link to SC 2.2.2 should use the full name (Success Criterion 2.2.2: Pause, Stop, Hide).

			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		function start() {
			updates = setInterval(change, 3000)
		}
	</script>
</body>
```

#### Failed Example 2

The text node automatically updates every 3 seconds after the page completes loading. The mechanism to stop the automatic updates is not visible.

```html
<body onload="start()">
	<p>Random number: <span id="target">1</span></p>
	<input type="button" onclick="stop()" value="Stop updates" style="position: absolute; top: -9999px; left: -9999px;" />

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		var updates
		function start() {
			updates = setInterval(change, 3000)
		}

		function stop() {
			clearInterval(updates)
		}
	</script>
</body>
```

#### Failed Example 3

The text node automatically updates every 3 seconds after the page completes loading. The mechanism to stop the automatic updates is not included in the accessibility tree.

```html
<body onload="start()">
	<p>Random number: <span id="target">1</span></p>

> 	<div onclick="stop()" aria-hidden="true">

Reviewer: This `div` should actually be a span to avoid the clickable area spreading over the full line.

Actually, the `div` and the `span` should be merged into a single element. There is nothing added to the example by having the `onclick` property on another node than the text.

		<span style="background-color:#333; color:white; padding:5px">Stop updates</span>
	</div>

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		var updates
		function start() {
			updates = setInterval(change, 3000)
		}

		function stop() {
			clearInterval(updates)
		}
	</script>
</body>
```

#### Failed Example 5

The text node automatically updates every 3 seconds after the page completes loading. The mechanism to stop the automatic updates has an accessible name that is only whitespace.

```html
<body onload="start()">
	<p>Random number: <span id="target">1</span></p>
	<div onclick="stop()" aria-label="   ">
		<span style="background-color:#333; color:white; padding:5px">Stop updates</span>
	</div>

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		var updates
		function start() {
			updates = setInterval(change, 3000)
		}

		function stop() {
			clearInterval(updates)
		}
	</script>
</body>
```

> #### Failed Example 6

Reviewer: Good job on that example 👍

		<style>
			.progress-wrap,
			.progress-bar {
				height: 20px;
				max-width: 300px;
				position: relative;
			}

			.progress-wrap {
				background: #eee;
			}

			.progress-bar {
				background: #cce0ff;
				width: 0;
				transition: width 1s;
			}

			.progress-text {
				position: absolute;
				top: 0;
				width: 100%;
				text-align: center;
			}

			.choices {
				margin: 3em;
			}
		</style>
	</head>

	<body onload="start()">
		<p>
			You can pick a price you like, or wait for the final.
		</p>
		<div class="progress-wrap">
			<div class="progress-bar" id="pb-demo"></div>
			<div class="progress-text" id="pb-text">0%</div>
		</div>
		<div class="choices">
			<button type="button" id="packa" onclick="select()">Pack A:</button>
		</div>
		<div>
			<p id="outcome"></p>
		</div>

		<script type="text/javascript">
			var n, updates, price

			function updatePB() {
				var bar = document.getElementById('pb-demo')
				var text = document.getElementById('pb-text')
				bar.style.width = n + '%'
				text.innerHTML = n + '%'
			}

			function updateNum() {
				var button = document.getElementById('packa')
				if (n < 100) {
					n = n + 10

> 					updatePB(n)

Reviewer: `updatePB` is using the global variable `n`. I'm in favour of adding a parameter to it (to reduce use of global variable), and thus keep this line as is, but currently it is not correct.

			target.innerText = number
		}

		var updates
		function start() {
			updates = setInterval(change, 3000)
		}

		function stop() {
			clearInterval(updates)
		}
	</script>
</body>
```

#### Passed Example 2

The text node automatically updates every 3 seconds after the page completes loading. A mechanism is available to pause and resume the automatic updates.

```html
<body onload="start()">
	<p>Random number: <span id="target">1</span></p>
	<input type="button" id="control" onclick="pause()" value="Pause updates" />

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		var updates
		var updating = false
		function start() {
			var control = document.getElementById('control')
			control.value = 'Pause updates'
			updating = true
			updates = setInterval(change, 3000)
		}

		function pause() {
			if (updating) {
				var control = document.getElementById('control')
				control.value = 'Resume updates'
				updating = false
				clearInterval(updates)
			} else {
				start()
			}
		}
	</script>
</body>
```

#### Passed Example 3

> The text node (part of a progress bar) automatically updates every second after the page completes loading. Since the auto-updating can be considered essential because it takes place during a loading phase when there can be no interaction, a mechanism to stop the auto-updating is not required and the example passes.

Author: It is correct. The progress loading starts after the page completes loading.

> The text node (part of a progress bar) automatically updates every second after the page completes loading. Since the auto-updating can be considered essential because it takes place during a loading phase when there can be no interaction, a mechanism to stop the auto-updating is not required and the example passes.

Author: I also prefer to leave it as an assumption. I'm updating the context of the example (to remove the essential nature of the update) and moving it to a failed example (because as you said, this fails the rule), so that we have an example with a progress bar.

```html
<html>
	<head>
		<title>Example page for rule efbfc7</title>
		<style>
			.progress-wrap,
			.progress-bar {
				height: 20px;
				max-width: 300px;
				position: relative;
			}
			.progress-wrap {
				background: #eee;
			}
			.progress-bar {
				background: #cce0ff;
				width: 0;
				transition: width 1s;
			}
			.progress-text {
				position: absolute;
				top: 0;
				width: 100%;
				text-align: center;
			}
			.choices {
				margin: 3em;
			}
		</style>
	</head>
	<body onload="start()">
		<p>
			Wait for the bar to reach 100% then make your choice
		</p>
		<div class="progress-wrap">
			<div class="progress-bar" id="pb-demo"></div>
			<div class="progress-text" id="pb-text">0%</div>
		</div>
		<div class="choices">
			<button type="button" id="packa" disabled>Pack A:</button>
			<button type="button" id="packb" disabled>Pack B:</button>
		</div>

		<script type="text/javascript">
			var n, updates

			function updatePB() {
				var bar = document.getElementById('pb-demo')
				var text = document.getElementById('pb-text')
				bar.style.width = n + '%'
				text.innerHTML = n + '%'
			}

			function updateNum() {
				if (n < 100) {
					n = n + 5
					updatePB(n)

> 					if (n % 10 === 0) {

Reviewer: This test is not really useful. We could as well change price every 5%.

						var buttonA = document.getElementById('packa')
						var buttonB = document.getElementById('packb')
						buttonA.innerHTML = 'Pack A: &euro;' + Math.floor(Math.random() * 1000)
						buttonB.innerHTML = 'Pack B: &euro;' + Math.floor(Math.random() * 1000)
					}
				} else {
					clearInterval(updates)
					var buttonA = document.getElementById('packa')

> 					var buttonB = document.getElementById('packb')

Reviewer: Given that these variables are used in both branches of the main test, they should be moved to the top-level of the function instead.

			target.innerText = number
		}

		function start() {
			updates = setInterval(change, 3000)
		}
	</script>
</body>
```

#### Failed Example 2

The text node automatically updates every 3 seconds after the page completes loading. The mechanism to stop the automatic updates is not visible.

```html
<body onload="start()">
	<p>Random number: <span id="target">1</span></p>
	<input type="button" onclick="stop()" value="Stop updates" style="position: absolute; top: -9999px; left: -9999px;" />

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		var updates
		function start() {
			updates = setInterval(change, 3000)
		}

		function stop() {
			clearInterval(updates)
		}
	</script>
</body>
```

#### Failed Example 3

The text node automatically updates every 3 seconds after the page completes loading. The mechanism to stop the automatic updates is not included in the accessibility tree.

```html
<body onload="start()">
	<p>Random number: <span id="target">1</span></p>
	<div onclick="stop()" aria-hidden="true">

> 		<span style="background-color:#333; color:white; padding:5px">Stop updates</span>

Author: I think so. It makes the mechanism sort of similar to a button.

> 		<span style="background-color:#333; color:white; padding:5px">Stop updates</span>

Reviewer: What I mean is that the example works the same way without the styling. The styling is only here to make the `span` look somewhat like a button, but nothing in the rule requires the mechanism to "look like a button" or whatever in this line. So, the exact same example without the styling is still failing for the exact same reasons.

That's why I think the styling should be removed to keep the example simple. When reading the example, I try to figure out what role the styling plays (with respect to this rule) and that create a bit of overhead on examples that are already complex (due to JS and all that).

	</div>

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		var updates
		function start() {
			updates = setInterval(change, 3000)
		}

		function stop() {
			clearInterval(updates)
		}
	</script>
</body>
```

#### Failed Example 4

The text node automatically updates every 3 seconds after the page completes loading. The mechanism to stop the automatic updates has an empty accessible name.

```html
<body onload="start()">
	<p>Random number: <span id="target">1</span></p>
	<div onclick="stop()" aria-label="">

> 		<span style="background-color:#333; color:white; padding:5px">Stop updates</span>

Author: Same as above

```html
<body onload="start()">
	<p>Random number: <span id="target">1</span></p>
	<div onclick="stop()" aria-label="   ">
		<span style="background-color:#333; color:white; padding:5px">Stop updates</span>
	</div>

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		var updates
		function start() {
			updates = setInterval(change, 3000)
		}

		function stop() {
			clearInterval(updates)
		}
	</script>
</body>
```

### Inapplicable

#### Inapplicable Example 1

The text node automatically updates every 3 seconds after the page completes loading but it is not visible.

```html
<body onload="start()">
	<p style="display: none">Random number: <span id="target">1</span></p>

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		function start() {
			updates = setInterval(change, 3000)
		}
	</script>
</body>
```

#### Inapplicable Example 2

> The text node automatically updates every 3 seconds but only as a result of an action from the user.

Reviewer: Is the pause mechanism important for this example? It seems that not auto-starting is enough for 2.2.2 to be happy.

---
id: efbfc7
name: Auto-updating text content can be paused or stopped
rule_type: atomic
description: |
  This rule checks that there are mechanisms to pause or stop the auto-updating of text content
accessibility_requirements: # Remove whatever is not applicable
  wcag20:2.2.2: # Pause, Stop, Hide (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
acknowledgements:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [visible][] [text node][] in the [flat tree][] of an [HTML document][] if:

> - the content of the [text node][] changes through an action that was not initiated by the user;

Reviewer 2: This seems ambiguous. what does it mean for a user to have initiated something? If I click a button that starts a timer for something to happen a week from now, is that event a week from now something I initiated? In that sense, isn't everything I do on my mac ultimately initiated by me?

Author: Rephrased it

> - the content of the [text node][] changes through an action that was not initiated by the user;

Reviewer 2: I don't think changing text nodes is a reliable way to determine there's a problem. For one, text can be added to the text node. That's a change that doesn't cause a problem. Or perhaps the current text is put into another text node, while the current text node is updated. This is pretty common stuff with virtual DOMs.

Another thing this doesn't consider is that the text node could be removed entirely. I wouldn't consider that a change, but it can be a failure.

Author: Given that the SC says "Auto-updating refers to content that updates or disappears" I'm updating the applicability to "changes, is replaced or removed"

- the change happens anytime after the [readiness][document readiness] of the [HTML document][] the [text node][] belongs to is equal to "complete"; and
- it is not the only [content][] in the [HTML document][].

## Expectation 1

> For the test target a [mechanism][] is provided to pause or stop the updating of the content of the [text node][].

Reviewer 2: The SC says hiding the updating content is always permitted.

**Note**: If there is more than one [text node][] with auto-updating content, a single [mechanism][] may be used to pause or stop updating all [text nodes][text node].

## Expectation 2

> The [mechanism][] to pause or stop the auto-updating of the content of the [text node][] is [visible][], has an [accessible name][] that is not empty (""), and is [included in the accessibility tree][].

Reviewer 2: This is not required by 2.2.2.

**Note**: This rule does not require the [mechanism][] to have a [label][] that is descriptive. Depending on how the [mechanism][] is implemented, a non-descriptive label would be a violation of another success criterion, like [Success Criterion 3.3.2: Labels or Instructions][sc 3.3.2]

## Assumptions

This rule assumes that the auto-updating of the content is not [essential][], which is listed as valid exception to [Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]. When the auto-updating of content is [essential][] this rule may produce incorrect results.

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [Understanding Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]
- [G186: Using a control in the Web page that stops moving, blinking, or auto-updating content][g186]
- [F16: Failure of Success Criterion 2.2.2 due to including scrolling content where movement is not essential to the activity without also including a mechanism to pause and restart the content][f16]

> ## Test Cases

Reviewer 2: All of these examples are inapplicable, because you are using innerText. `innerText` doesn't change the text node, it replaces it.

Author: This is addressed by the new applicability

	<input type="button" onclick="stop()" value="Stop updates" />

	<script type="text/javascript">
		function change() {
			var target = document.getElementById('target')
			var number = Math.floor(Math.random() * 1000)
			target.innerText = number
		}

		var updates
		function start() {
			updates = setInterval(change, 3000)
		}

		function stop() {
			clearInterval(updates)
		}
	</script>
</body>
```

[accessible name]: #accessible-name 'Definition of accessible name'
[ascii whitespace]: https://infra.spec.whatwg.org/#ascii-whitespace
[child]: https://dom.spec.whatwg.org/#concept-tree-child
[content]: https://www.w3.org/TR/WCAG21/#dfn-content
[document readiness]: https://www.w3.org/TR/html53/dom.html#current-document-readiness
[essential]: https://www.w3.org/WAI/WCAG21/Understanding/pause-stop-hide.html#dfn-essential
[flat tree]: https://drafts.csswg.org/css-scoping/#flat-tree
[f16]: https://www.w3.org/WAI/WCAG21/Techniques/failures/F16
[g186]: https://www.w3.org/WAI/WCAG21/Techniques/general/G186
[html document]: https://dom.spec.whatwg.org/#html-document
[included in the accessibility tree]: #included-in-the-accessibility-tree 'Definition of included in the accessibility tree'

> [label]: https://www.w3.org/TR/WCAG21/#dfn-labels

Reviewer 2: This definition is ambiguous. I don't think we can use this.

> [mechanism]: https://www.w3.org/TR/WCAG21/#dfn-mechanism

Reviewer 2: This definition is ambiguous. I don't think we can use this.

Author: It is being used in [audio or video that plays automatically has a control mechanism](https://act-rules.github.io/rules/4c31df)

Author: Not being used anymore in this rule

    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgements:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [visible][] [text node][] of an [HTML document][] if:

> - the content of the [text node][] either changes, is replaced or removed without any user intervention;

Author: I'll get rid of text nodes in the applicability and propose a definition of user intervention.

input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgments:
  authors:
    - Carlos Duarte
---

## Applicability

The rule applies to any [visible][] [HTML element][] if:

- **(changed)** the `innerText` property of the [element][html element] changes; and
- **(no child changed)** the [element][html element] does not have [children][child] whose `innerText` property also changes; and

> - **(not alone)** the [element][html element] has [sibling][] or [ancestor][] elements with non-empty `innerText` properties; and

Reviewer: Wouldn't that mean that if the only other content is not text (image, video, …) there is no problem?

Hum, that would only make the rule be inapplicable "too often", so not a big deal anyway…

Author: To keep the rule unambiguous I'd leave it as is, risking some inapplicable cases, as you said.
However, re-reading this, I found that the `innerText` of the ancestors needs to be different from the `innerText` of the target element, otherwise something like the example below would be applicable because the outer `<div>` has inner text.
```html
<div>
  <div>auto-updating text</div>
</div>
```

> - **(not alone)** the [element][html element] has [sibling][] or [ancestor][] elements with non-empty `innerText` properties; and

Reviewer: I guess for ancestors we also need the non-empty `innerText` to be different from the `innerText` of the changing element.

- the [document][html document] the [element][html element] belongs to:
  - **(ready)** has [readiness][document readiness] equal to "complete" before the change happens; and
  - **(inactive)** is associated with a [window][] that has the [last activation timestamp][] equal to "positive infinity".

## Expectation

> For the test target there exists an [instrument][] in the same [HTML document][] of the test target, or in another [HTML document][] that is [linked][hyperlink] from the [HTML document][] of the test target, to achieve one of the following objectives:

Reviewer: Do we want to require conditions on the link (visible, accessible, …) or not?

Author: There is nothing in SC 2.2.2 about that. In other rules where that was the case, and where I had included similar requirements in the expectation, I had to take them out following reviewers' requests. That's why I left those out in this one.

## Applicability

> The rule applies to any [HTML element][] that has a descendant in the [flat tree][] if:

Reviewer 2: ```suggestion
The rule applies to any [HTML element][] that has a [visible][] [text node][] as a [descendant][] in the [flat tree][] if:
```

## Applicability

The rule applies to any [visible][] [HTML element][] in an [HTML document][] if:

> - the `innerText` property of the [element][html element] is updated with a determinable frequency;

Reviewer: I think "updated" is not precise enough. For example, it is not clear whether "deleting" is "updating". Or whether overwriting with the same value is "updating". Maybe going in the direction of "changed" is more precise 🤔.

Author: Agreed and reverted back to "changes" instead of "updates".

> - the `innerText` property of the [element][html element] is updated with a determinable frequency;

Author: I was also concerned about the frequency being undecidable (not about the determinable). Removed it from the new version.

- the [element][html element] does not have [children][child] whose `innerText` property is also updated;

> - the updating starts when the [readiness][document readiness] of the [HTML document][] the [element][html element] belongs to is equal to "complete";

Author: Replaced "starts" with "happens".

## Applicability

The rule applies to any [HTML element][] with a [visible][] [text node][] if:

- **changed:** The `innerText` property of the [element][html element] changes multiple times within a 10 minute time span where there is no [user interaction][]; and
- **no child changed:** the [element][html element] does not have [children][child] whose `innerText` property also changes; and
- **not alone:** the [element][html element] has an [ancestor][] element with a non-empty `innerText` property whose value is different from the `innerText` of the test target.

**Note:** The test targets for this rule are elements that have their text content change multiple times, on a page with further content, if that change happens without user intervention.

> **Note:** The 10 minute time span is arbitrary. Content that updates less frequently, or that doesn't update on a regular interval, may still not satisfy the success criteria, but testers generally don't wait hours to see if a page changes.

Reviewer: ```suggestion
**Note:** The 10 minute time span is an arbitrary limit which is not included in WCAG. Content that updates less frequently, or that doesn't update on a regular interval, will be inapplicable for this rule but 
may nonetheless fail [Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]. It is however a much lesser accessibility issue, and without any time limit this rule would be virtually untestable.
```

## Expectation

For each test target there exists an [instrument][] in the same [web page](#web-page-html), or in another [web page](#web-page-html) that is [linked][hyperlink] from the [web page](#web-page-html) of the test target, to achieve one of the following objectives:

- pause and resume the change of the [visible text content][]; or
- stop the change of the [visible text content][]; or
- hide the content that changes; or
- alter the frequency of the changes of the [visible text content][].

> **Note:** If there is more than one test target, a single [instrument][] to pause, stop, hide or alter the frequency for all test targets is sufficient.

Reviewer: ```suggestion
**Note:** If there is more than one test target, the same [instrument][] may be used to pause, stop, hide or alter the frequency of several or even all test targets.
```
Making it clear that it is not "one or all" but "any number".

## Assumptions

- This rule assumes that the auto-updating of the content is not [essential][], which is listed as valid exception to [Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]. When the auto-updating of content is [essential][] this rule may produce incorrect results.
- This rule assumes that any [content][] changes are enabled by the content of [HTML document][] the test target belongs to. Changes originated by any other sources (e.g. browser shortcuts, browser extensions, browser settings, user agents, external browser applications) are not considered.

> - This rule assumes that all user actions are transmitted by the user agent to the [HTML document][]. If there are other event sources that result from a user action this rule might fail but the success criteria might still be met.

Reviewer: ```suggestion
- This rule assumes that all user actions are transmitted by the user agent to the [HTML document][]. If there are other event sources that result from a user action this rule might fail but the success criterion might still pass.
```

> - This rule assumes that available mechanisms for controlling the update of content rely on [activation][]. If there are other mechanisms that do not really on [activation][] the rule might fail but the success criteria might still be met.

Reviewer: ```suggestion
- This rule assumes that available mechanisms for controlling the update of content rely on [activation][]. If there are other mechanisms that do not really on [activation][] the rule might fail but the success criterion might still pass.
```

The rule applies to any [visible][] [HTML element][] in an [HTML document][] if:

> - the `innerText` property of the [element][html element] changes without the user [activating][activate] or [focusing][focus] any [element][html element] in the same [HTML document][];

Reviewer: I like the way this goes 😄 

* I'm not entirely sure this takes care of the delay between activation and change (if I click a button and the page changes 2h later, is it OK?) Maybe add a condition like "the window has [Transient activation](https://html.spec.whatwg.org/#transient-activation)" to put a reasonable time limit which is (AFAICT) UA dependant, hence we're not the ones deciding what is "reacting to user input".
* Can I unfocus an element without focusing another one? If so, that might be a user activation which can change page and would be good for WCAG but not for this rule. Maybe refer to [Activation triggering input event](https://html.spec.whatwg.org/#activation-triggering-input-event) for an "official" list of what is user interaction.
* Similar question with global events that are not really related to any element but can still change stuff.

Also, wondering whether there should be an assumption on only looking at activations enabled by content (and not, say, browser shortcuts or extensions), _à la_ https://github.com/act-rules/act-rules.github.io/pull/873/files#diff-080b736abdfda182cd29c13427ee2aecR34

Author: You're focusing on changes that result from user actions. The rule applies to changes that do not result from user actions. If the user did something that triggered an update that only happens 2 hours in the future then that is not applicable (I'm not saying it's okay, but I don't think we can go down that path without compromising the objectivity so I argue that we should stick with an applicability that does not consider "delayed updates").

Good comment about the assumption. I'll add it.

Reviewer: Hmm, so you're saying that a delayed activation is considered as a user activation and that will make the rule not fail but the SC fail, which is OK.

I think you have a point…

I was suggesting to use the Transient activation state as it is an objective and unambiguous state that UA consider are "close enough to user action that we can still do stuff". But it might indeed get too hairy.

Hum, no, you're right, we can't really put a delay because something like Inapplicable Example 2 (which is OK) has updates that happen as long after the click as we want (they never stop).

---
I still think there is an issue on what is defined as "user interaction".
Typically, Inapplicable Example 2 is good because the updates only start after I click (and never stop).

What if there is a page with a global event listener (not linked to any element). When I press `ctrl+s`, the update start. This is not the user activating or focusing any element, so the rule is applicable. There is no component to stop the update, so the rule fail. But SC 2.2.2 does not fail because the update does not "start automatically". Failing the rule without failing the SC is bad…

> The rule applies to any [visible][] [HTML element][] in an [HTML document][] if:

Reviewer: Suggesting reformulating the enumeration (splitting conditions on the element and on the changes, and naming conditions as Wilco is introducing in https://github.com/act-rules/act-rules.github.io/pull/1171/files#diff-ba71ed44d1a01755345d5f6362432855R34-R39) in the line of:

The rule applies to any element in a document if the element:
- **(not alone)** is not the only content; and
- **(change)** has a `innerText` that changes; and
- **(no child change)** does not have children …; and
- is part of a document:
  - **(ready)** whose readiness is complete when the change happens; and
  - **(inactive)** is associated to a window with infitine timestamp.

I think at least the nested list is good to avoid the complex reverse relatives to qualify the document or the window.

- the `innerText` property of the [element][html element] changes;
- the [element][html element] does not have [children][child] whose `innerText` property is also changed;
- the change happens after the [readiness][document readiness] of the [HTML document][] the [element][html element] belongs to is equal to "complete";
- the change happens when the [last activation timestamp][] of the [window][] the [HTML document][] is associated with has the value "positive infinity";
- it is not the only [content][] in the [HTML document][].

## Expectation 1

> For the test target a [user interface component][] is provided to pause, stop, hide or alter the frequency of the changes of the [visible text content][].

Reviewer: The enumeration does not read very well due to the alter part having a specific part ("the frequency"). Makes it look like we could "stop the frequency" which does not make sense… (and "hide" refers to the content, while "pause" and "stop" refer to the changes and "alter" to "the frequency of the changes").

Also, I would replace "is provide" by a "there exists" formulation.

**Note:** The 10 minute time span is an arbitrary limit which is not included in WCAG. Content that updates less frequently, or that doesn't update on a regular interval, will be inapplicable for this rule but may nonetheless fail [Success Criterion 2.2.2: Pause, Stop, Hide][sc 2.2.2]. It is however a much lesser accessibility issue, and without any time limit this rule would be virtually untestable.

## Expectation 1

> For each test target there exists an [instrument][] to achieve one of the following objectives:

Reviewer: I think we need to go the "set of instruments" path here also.
There could be one button for pausing and a different one for resuming.

Author: I agree. The new example makes that clear :-)

## Expectation 2

> The set of [instruments][instrument] from Expectation 1 is in one of the following:

Reviewer: Do you mean that all instruments in the set are in the same place, or can they be in different places?

What if the "pause" button is on the web page but the "resume" button is on a modal?

#### Inapplicable Example 2

> The text content automatically updates but only as a result of the user activating an element on the page.

Reviewer: ```suggestion
The text content automatically updates but only as a result of the user [activating][activate] an element on the page.
```
